tests/unittests/test_runs/test_fallback_datasource.py

```diff
@@ -0,0 +1,240 @@
+import os
+import re
+import json
+from cloudinit.sources import DataSourceConfigDrive as dscd
+from cloudinit.sources import DataSourceOpenStack as dsos
+from cloudinit.sources import DataSourceOVF as dsovf
+from cloudinit import helpers
+from cloudinit.tests import helpers as tests_helpers
+from cloudinit.tests.helpers import mock
+from cloudinit import util
+from cloudinit.tests.helpers import populate_dir
+from cloudinit.sources.helpers import openstack
+from cloudinit import stages
+from cloudinit import safeyaml
+import httpretty as hp
+from io import StringIO
+from urllib.parse import urlparse
+
+
+OSTACK_META = {
+    'availability_zone': 'nova',
+    'hostname': 'sm-foo-test.novalocal',
+    'meta': {'dsmode': 'local', 'my-meta': 'my-value'},
+    'name': 'sm-foo-test',
+    'uuid': 'b0fa911b-69d4-4476-bbe2-1c92bff6535c'}
+
+
+OS_FILES = {
+    'openstack/latest/meta_data.json': json.dumps(OSTACK_META),
+    'openstack/latest/network_data.json': json.dumps(
+        {'links': [], 'networks': [], 'services': []}),
+}
+
+
+OVF_ENV_CONTENT = """<?xml version="1.0" encoding="UTF-8"?>
+<Environment xmlns="http://schemas.dmtf.org/ovf/environment/1"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns:oe="http://schemas.dmtf.org/ovf/environment/1"
+  xsi:schemaLocation="http://schemas.dmtf.org/ovf/environment/1 ../dsp8027.xsd"
+  oe:id="WebTier">
+  <!-- Information about hypervisor platform -->
+  <oe:PlatformSection>
+      <Kind>ESX Server</Kind>
+      <Version>3.0.1</Version>
+      <Vendor>VMware, Inc.</Vendor>
+      <Locale>en_US</Locale>
+  </oe:PlatformSection>
+  <!--- Properties defined for this virtual machine -->
+  <PropertySection>
+{properties}
+  </PropertySection>
+</Environment>
+"""
+
+CFG_DRIVE_FILES_V2 = {
+    'openstack/latest/meta_data.json': json.dumps(OSTACK_META)}
+
+BASE_URL = "http://169.254.169.254"
+
+EC2_VERSIONS = [
+    'latest',
+]
+
+
+def fill_properties(props, template=OVF_ENV_CONTENT):
+    lines = []
+    prop_tmpl = '<Property oe:key="{key}" oe:value="{val}"/>'
+    for key, val in props.items():
+        lines.append(prop_tmpl.format(key=key, val=val))
+    indent = "        "
+    properties = ''.join([indent + line + "\n" for line in lines])
+    return template.format(properties=properties)
+
+
+def _register_uris(version, ec2_files, ec2_meta, os_files):
```

Comment:
Does this differ from what's in tests/unittests/test_datasource/test_openstack?  If not, let's import this method from there; if it does, please explain what's different.

Reply:
it doesn't differ. Addressed in latest commit